### PR TITLE
fix(enricher): retain image names for delegated scans

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

Previously, there was a divergence between how the cached, existing image's values were enriched during scans. For delegated scanning, it wouldn't include the merged version of iamge scans, for non-delegated scans it would include them.

This would lead to a difference in behavior based on the scan "method" used. Culprit was the current call of `updateImageFromDatabase` vs `fetchFromDatabase`:
`fetchFromDatabase` retains image names only for `UseImageNamesRefetchCachedValues`, whereas `updateImageFromDatabase` does so for all enrichment options.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

- see added unit tests.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* master
  * **PR #8428** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/8428" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/8428?utm_source=stack-comment).

File: pkg/images/enricher/enricher_impl.go [hunk 1]
```diff
@@ -160,11 +160,7 @@ func (e *enricherImpl) delegateEnrichImage(ctx context.Context, enrichCtx Enrich
 	// or forcing re-scan will trigger updates as necessary.
 	existingImg, exists := e.fetchFromDatabase(ctx, image, enrichCtx.FetchOpt)
 	if exists && cachedImageIsValid(existingImg) {
-		image.Metadata = existingImg.GetMetadata()
-		image.Scan = existingImg.GetScan()
-		image.Signature = existingImg.GetSignature()
-		image.SignatureVerificationData = existingImg.GetSignatureVerificationData()
-		image.Notes = existingImg.GetNotes()
+		e.updateImageWithExistingImage(image, existingImg, enrichCtx.FetchOpt)
```
Comment (dcaravel) [inline pkg/images/enricher/enricher_impl.go:163]: Believe as is the return from `roxctl image scan` would never return a result if a DB entry exists (unless we change handling of `IgnoreExistingImages` mentioned above), or do something like:

```suggestion
		updated := e.updateImageWithExistingImage(image, existingImg, enrichCtx.FetchOpt)
		if updated {
			e.cvesSuppressor.EnrichImageWithSuppressedCVEs(image)
			e.cvesSuppressorV2.EnrichImageWithSuppressedCVEs(image)

			log.Debugf("Delegated enrichment returning cached image for %q", image.GetName().GetFullName())
			return true, nil
		}
```

**Review by dcaravel — CHANGES_REQUESTED** [review]
There appear to be multiple places this issue can occur:
1. roxctl non-delegated losing `names`
2. roxct delegated losing `names` (this PR)
3. Sensor delegated losing `names` (`ROX-19095` to address this)

Regarding the first item, the root cause appears to be handling of `IgnoreExistingImages`, which is [set by default](https://github.com/stackrox/stackrox/blob/204b87e6f6f79661b52650467f7821ea08f93f25/central/image/service/service_impl.go#L307) in `ScanImage` requests (ie: `roxctl image scan`)

The outcome is strange with the current usage of `IgnoreExistingImages` - by default `ScanImage` ignores all the cache, whereas the `--force` flag ignores only some of the cache - I would assume the reverse be true, using `--force` should ignore more of the cache (ie: should be used in scenarios where the cache is stale/inaccurate to 'reset') - looks like the behavior changed in [this PR](https://github.com/stackrox/stackrox/pull/5339), which I believe we should revisit.

The outcome when using / not using the `force` flag while delegated a scan seems to be more logically accurate currently (unless I'm missing something).  I'd rather add populating `image.Names` with the value from the database the address this issue specific to delegated scanning instead or revisit the handling of `IgnoreExistingImages` for all scans.

File: pkg/images/enricher/enricher_impl.go (hunk 1 above)
Comment (dhaus67) [reply pkg/images/enricher/enricher_impl.go:163]: Fixed this.


On the notion of `IgnoreExistingImages` and its usage in the `ScanImage` API:
I feel like the _behavior_ of `IgnoreExistingImages` is IMO fine, where I think the issue lies is that `ScanImage` uses it in the first place. Looking at the history of the code, this seems to have been in place for quite some time (multiple years back).

Within the usage of roxctl, we most of the time advertise to use `roxctl image check`, which does scanning + policy checking. For the detection service API, the default fetch option is `UseCachesIfPossible`. This also totally makes sense, given that for this API a _force_  option is there as well.

I'll file a follow-up PR to change the `ScanImage` API to use `UseCachesIfPossible` as default, the behavior of `IgnoreExistingImages`, specifically the use-case outline with the watched images, makes sense for me, just not for the scanning API to use.

**Review by dcaravel — APPROVED** [review]
LGTM!

**Pull request merged**